### ISSUE TITLE
lxc copy to remote : Error transferring container data: websocket: bad handshake

### ISSUE THREAD

**Issue by arsiesys** [lxc/lxd#1379]
Hello

When I'm trying to copy a container to a remote LXD server, I got this error : 

lxc copy znc srv01:
error: Error transferring container data: websocket: bad handshake

If I put on the monitor on the remote server,
I saw this part : 
```
metadata:
  context: {}
  level: dbug
  message: |
    Raw response: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
    <html><head>
    <title>404 Not Found</title>
    </head><body>
    <h1>Not Found</h1>
    <p>The requested URL /1.0/operations/8463b376-1320-4009-926d-535ad7cca769/websocket was not found on this server.</p>
    <hr>
    <address>Apache/2.2.22 (Debian) Server at XX.XX.XX.XX Port 443</address>
    </body></html>
timestamp: 2015-12-08T21:53:47.994986195+01:00
type: logging
```

lxc copy debug output : 
```
DBUG[12-08|21:01:31] Fingering the daemon
DBUG[12-08|21:01:31] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"api_compat":1,"auth":"trusted","config":{"core.https_address":"xx.xx.xx.xx","core.trust_password":true,"images.remote_cache_expiry":"10"},"environment":{"addresses":["xx.xx.xx.xx"],"architectures":[2,1],"driver":"lxc","driver_version":"1.1.5","kernel":"Linux","kernel_architecture":"x86_64","kernel_version":"3.13.0-71-generic","server":"lxd","server_pid":1133,"server_version":"0.23","storage":"btrfs","storage_version":"3.12"}},"operation":""}

DBUG[12-08|21:01:31] Pong received
DBUG[12-08|21:01:31] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"architecture":2,"config":{"volatile.base_image":"ad2b9a455590aed7503eb942ed10687c92b56b701d06f72efd2af1263779b489","volatile.eth0.hwaddr":"00:16:3e:c0:e9:7e","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":165536,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":165536,\"Nsid\":0,\"Maprange\":65536}]"},"devices":{},"ephemeral":false,"expanded_config":{"volatile.base_image":"ad2b9a455590aed7503eb942ed10687c92b56b701d06f72efd2af1263779b489","volatile.eth0.hwaddr":"00:16:3e:c0:e9:7e","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":165536,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":165536,\"Nsid\":0,\"Maprange\":65536}]"},"expanded_devices":{"eth0":{"hwaddr":"00:16:3e:c0:e9:7e","nictype":"bridged","parent":"lxcbr0","type":"nic"}},"name":"znc","profiles":["default"],"status":{"status":"Running","status_code":103,"init":4039,"processcount":7,"ips":[{"interface":"eth0","protocol":"IPV4","address":"10.0.3.84","host_veth":""},{"interface":"lo","protocol":"IPV4","address":"127.0.0.1","host_veth":""},{"interface":"lo","protocol":"IPV6","address":"::1","host_veth":""}]}},"operation":""}

DBUG[12-08|21:01:31] Fingering the daemon
DBUG[12-08|21:01:32] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"api_compat":1,"auth":"trusted","config":{"core.https_address":"vv.vv.vv.vv","core.trust_password":true,"images.remote_cache_expiry":"10"},"environment":{"addresses":["vv.vv.vv.vv"],"architectures":[2,1],"driver":"lxc","driver_version":"1.1.5","kernel":"Linux","kernel_architecture":"x86_64","kernel_version":"3.13.0-71-generic","server":"lxd","server_pid":7277,"server_version":"0.23","storage":"btrfs","storage_version":"3.12"}},"operation":""}

DBUG[12-08|21:01:32] Pong received
DBUG[12-08|21:01:32] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":["/1.0/profiles/default","/1.0/profiles/migratable","/1.0/profiles/priv"],"operation":""}

DBUG[12-08|21:01:32] Posting {"migration":true}
 to http://unix.socket/1.0/containers/znc
DBUG[12-08|21:01:32] Raw response: {"type":"async","status":"OK","status_code":100,"metadata":{"id":"07e7a7c7-d8bb-4620-b8dc-3b33e04bf5ea","class":"websocket","created_at":"2015-12-08T21:01:32.267358938Z","updated_at":"2015-12-08T21:01:32.267358938Z","status":"Running","status_code":103,"resources":{"containers":["/1.0/containers/znc"]},"metadata":{"control":"87f12090b2101af90b64a9d53c94c95a008fd08111c3da4509c74fb648d9334c","criu":"64194ea7558a3d179a8ba030e09640560e40ca540dc6a667df0bf661e5756dda","fs":"c1c1c1bfb6d7d2c430b2f9a464712f159b6e4e00a03c3ad15af634695514c0a8"},"may_cancel":false,"err":""},"operation":"/1.0/operations/07e7a7c7-d8bb-4620-b8dc-3b33e04bf5ea"}

DBUG[12-08|21:01:32] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"api_compat":1,"auth":"trusted","config":{"core.https_address":"xx.xx.xx.xx","core.trust_password":true,"images.remote_cache_expiry":"10"},"environment":{"addresses":["xx.xx.xx.xx"],"architectures":[2,1],"driver":"lxc","driver_version":"1.1.5","kernel":"Linux","kernel_architecture":"x86_64","kernel_version":"3.13.0-71-generic","server":"lxd","server_pid":1133,"server_version":"0.23","storage":"btrfs","storage_version":"3.12"}},"operation":""}

DBUG[12-08|21:01:32] Posting {"architecture":2,"config":{},"devices":{},"ephemeral":false,"name":"znc","profiles":["default"],"source":{"base-image":"ad2b9a455590aed7503eb942ed10687c92b56b701d06f72efd2af1263779b489","mode":"pull","operation":"wss://xx.xx.xx.xx/1.0/operations/07e7a7c7-d8bb-4620-b8dc-3b33e04bf5ea/websocket","secrets":{"control":"87f12090b2101af90b64a9d53c94c95a008fd08111c3da4509c74fb648d9334c","criu":"64194ea7558a3d179a8ba030e09640560e40ca540dc6a667df0bf661e5756dda","fs":"c1c1c1bfb6d7d2c430b2f9a464712f159b6e4e00a03c3ad15af634695514c0a8"},"type":"migration"}}
 to https://91.121.196.236:8443/1.0/containers
DBUG[12-08|21:01:32] Raw response: {"type":"async","status":"OK","status_code":100,"metadata":{"id":"963cfab0-d4fa-4744-b143-5d00dd2822f9","class":"task","created_at":"2015-12-08T22:01:15.920453745+01:00","updated_at":"2015-12-08T22:01:15.920453745+01:00","status":"Running","status_code":103,"resources":{"containers":["/1.0/containers/znc"]},"metadata":null,"may_cancel":false,"err":""},"operation":"/1.0/operations/963cfab0-d4fa-4744-b143-5d00dd2822f9"}

DBUG[12-08|21:01:32] 1.0/operations/963cfab0-d4fa-4744-b143-5d00dd2822f9/wait
DBUG[12-08|21:01:33] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"id":"963cfab0-d4fa-4744-b143-5d00dd2822f9","class":"task","created_at":"2015-12-08T22:01:15.920453745+01:00","updated_at":"2015-12-08T22:01:15.920453745+01:00","status":"Failure","status_code":400,"resources":{"containers":["/1.0/containers/znc"]},"metadata":null,"may_cancel":false,"err":"Error transferring container data: websocket: bad handshake"},"operation":""}

error: Error transferring container data: websocket: bad handshake
```

It seems that it try to join the port 443 instead of 8443 configured.

Current LXC version on both server : 0.23
OS  on both server: Ubuntu server 14.04
Container storage are on a BTRFS fs on both server

Please ask for missing and needed log and I will try to provide them.

**Comment by stgraber** [comment]
Can you paste:
 - lxc remote list
 - lxc info srv01:

**Comment by arsiesys** [comment]
lxc remote list
+--------+-----------------------------------------+--------+
|  NAME  |                   URL                   | PUBLIC |
+--------+-----------------------------------------+--------+
| images | https://images.linuxcontainers.org:8443 | YES    |
| local  | unix://                                 | NO     |
| srv01  | https://vv.vv.vv.vv:8443             | NO     |
+--------+-----------------------------------------+--------+

lxc info srv01:
apicompat: 1
auth: trusted
environment:
  addresses:
  - vv.vv.vv.vv
  architectures:
  - 2
  - 1
  driver: lxc
  driverversion: 1.1.5
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 3.13.0-71-generic
  server: lxd
  serverpid: 7277
  serverversion: "0.23"
  storage: btrfs
  storageversion: "3.12"
config:
  core.https_address: vv.vv.vv.vv
  core.trust_password: true
  images.remote_cache_expiry: "10"
public: false

**Comment by stgraber** [comment]
Trying to reproduce this but so far hitting other bugs along the way :)

**Comment by arsiesys** [comment]
I put a workaround to temporary resolve the issue. I used iptables to nat the 443 to 8443 when packets come from the remote lxd server. Now I have other issue... ah ah ^^'.
1/ Even if migratable profile is enable (and container restarted), I got some criu error.
2/ If I shutdown the container before copy (no crui issue), this time, it's rsync which seems to start 5s and do not progress (perhaps network iptables configuration from my side, need to check). I have to kill the rsync and nc process.

But, I will make ticket about that after this current issue :p.

So until resolution, best way to migrate (without live) for me : 
- lxc init a container on the remote server with a random image
- rsync my container data
- apply correct chown
- start and have fun

Thanks for your work on LXD ;). I hope LXD will meet success. I definitely adopted it (bug included :D).
